lib/components/user/favorite-location.js

```diff
@@ -0,0 +1,236 @@
+      handler = this._handleLocationChange
+      hideDelete = false // isFixed && isBlank(location.address)
+      placeholder = isFixed ? `Add ${location.type}` : 'Enter a favorite address'
+      DropdownBtn = StyledDropdown
+    }
+
+    // Show a dropdown for editing the icon, unless isFixed = true (for 'home' and 'work' locations).
+    // The dropdown has a predefined list of items for location types.
+    const iconControl = isFixed
+      ? (
+        <StyledAddon title={iconTitle}>
+          <FontAwesome name={icon} />
+        </StyledAddon>
+      )
+      : (
+        <LocationTypeSelector
+          DropdownButtonComponent={DropdownBtn}
+          id={`icon-select-${index}`}
+          onChange={this._handleLocationTypeChange}
+          selectedType={location ? location.type : 'custom'}
+        />
+      )
+
+    return (
+      <FormGroup>
+        <InputGroup>
+          {iconControl}
+          {/* wrapper with z-index override needed for showing location menu on top of other controls. */}
+          <span className='form-control' style={{ padding: 0, zIndex: 'inherit' }}>
+            <FavoriteLocationField
+              inputPlaceholder={placeholder}
+              location={location}
+              locationType='to'
+              onLocationSelected={handler}
+              showClearButton={false}
+            />
+          </span>
+
+          {/* FIXME: hide the delete button for 'home' and 'work' after user clicks it
+              (there are challenging layouts).
+            */}
+          {!hideDelete && (
+            <InputGroup.Button>
+              <Button
+                aria-label='Delete this location'
+                onClick={this._handleDelete}
+                title='Delete this location'
+              >
+                <Glyphicon glyph='trash' />
+              </Button>
+            </InputGroup.Button>
+          )}
+        </InputGroup>
+      </FormGroup>
+    )
+  }
+}
+
+const customLocationType = {
+  icon: 'map-marker',
+  text: 'Custom',
+  type: 'custom'
+}
+
+const locationTypes = [
+  // TODO: add more non-home/work types
+  {
+    icon: 'cutlery',
+    text: 'Dining',
+    type: 'dining'
+  },
+  customLocationType
+]
+
+/**
+ * Displays a dropdown for selecting one of multiple location types.
+ */
+const LocationTypeSelector = ({ DropdownButtonComponent, id, onChange, selectedType }) => {
+  // Fall back to the 'custom' icon if the desired type is not found.
+  const locationType = locationTypes.find(t => t.type === selectedType) || customLocationType
+
+  return (
+    <DropdownButtonComponent
+      componentClass={InputGroup.Button}
+      id={id}
+      title={<FontAwesome name={locationType.icon} style={{width: '10px'}} />}
+    >
+      {locationTypes.map((t, index) => (
+        <MenuItem eventKey={t} key={index} onSelect={onChange}>
+          <FontAwesome name={t.icon} style={{width: '15px'}} /> {t.text}
+        </MenuItem>
+      ))}
+    </DropdownButtonComponent>
+  )
+}
```

Comment:
This file is big enough that it might be worth moving this selector to its own file.

Reply:
Updated in https://github.com/opentripplanner/otp-react-redux/pull/280/commits/9298f94739c2196c0a6f47717a9481c12ad1e306.